### DISCUSSION TITLE
Running a package installed with npm in a local repo

### DISCUSSION THREAD

**Discussion by anders-vandvik** [jdx/mise#1551] [Q&A]
Hi!

I am just getting started with `mise`.

I have a Python and Node project for which I'd like to use `mise` to maintain the runtimes. I have defined the .mise.toml as follows
````
[tools]
python = {version="3.11", virtualenv=".venv"}
node = "lts"
````
I have run `mise install` when cd'd into the repo, and that seems to run fine. 

`which python` returns /Users/me/repositories/my-repo/.venv/bin/python
`which node` returns /Users/me/.local/share/mise/installs/node/lts/bin/node
`which npm` returns /Users/me/.local/share/mise/installs/node/lts/bin/npm

I have the following dependencies in the `packages.json`
````
{
  "dependencies": {
    "@apidevtools/swagger-cli": "latest",
    "@openapitools/openapi-generator-cli": "^2.9.0"
  }
}
````
I installed these with `npm install`. `npm list` returns
````
my-repo@ /Users/me/repositories/my-repo/MLF-Sales-Trainer
├── @apidevtools/swagger-cli@4.0.4
└── @openapitools/openapi-generator-cli@2.9.0
````
When I try to run a scrip that uses `swagger-cli` and `openapi-generator-cli`, the commands cannot be found. 
The `node` part of my `PATH` variable shows /Users/me/.local/share/mise/installs/node/lts/bin

I suspect this is related to PATH and the fact that it does not point to the local node version that's in `my-repo`. 
I would really appreciate some help here!
What am I doing wrong? How do I fix this? How do I make sure that whenever I cd into a repo that has a local node version, the `PATH` points to that node version?

**Comment by Marukome0743** [comment]
The selected Node installation is working correctly here. Packages installed by a project-local `npm install` do not place their executables in the Node installation's own `bin` directory. npm places them in:

```text
<project>/node_modules/.bin
```

`npm run` automatically adds this directory to `PATH`, and `npx` can also run a locally installed package. For example:

```sh
npm run generate
npx swagger-cli ...
```

If you want commands such as `swagger-cli` to be directly available in the activated mise environment, add the local bin directory explicitly:

```toml
[tools]
node = "lts"

[env]
_.path = "./node_modules/.bin"
```

The relative path is resolved from the directory containing `mise.toml`, so it continues to work when entering a subdirectory of the project.

I verified this against the current main branch with Node 22.23.2. I installed `cowsay@1.6.0` locally with npm, added the `_.path` entry above, and successfully invoked `cowsay` directly through `mise exec` without using its full path.

Therefore, this is separate from whether mise selected the correct local Node version: mise manages the runtime, while npm manages the project's `node_modules/.bin` executables.

*AI-assisted — Tool: Codex; model: OpenAI/GPT-5; version: unavailable.*
